File: src/main/java/sample/design_pattern/Mediator.java
/*
 * Copyright 2013 SatoShun
 *
 * Mediator Pattern
 */

package sample.design_pattern;

class Game {
    private IPlayer player;
    private IPlayer enemy;

    private int count;

    public Game() {
        count = 0;
    }

    public void setPlayer(IPlayer player) {
        this.player = player;
    }

    public void setEnemy(IPlayer enemy) {
        this.enemy = enemy;
    }

    public void change(IPlayer actionPlayer) {
        if (actionPlayer instanceof Player) {
            count += 1;
        } else if (actionPlayer instanceof Enemy) {
            count -= 1;
        }
    }

    public int getCount() {
        return count;
    }
}


interface IPlayer {
    void win();
}


class Player implements IPlayer {
    private Game game;

    public Player(Game game) {
        this.game = game;
    }

    public void win() {
        game.change(this);
    }
}


class Enemy implements IPlayer {
    private Game game;

    public Enemy(Game game) {
        this.game = game;
    }

    public void win() {
        game.change(this);
    }
}


class Mediator {
    public static void main(String[] args) {
        Game game = new Game();
        IPlayer player = new Player(game);
        IPlayer enemy = new Enemy(game);

        enemy.win();
        System.out.println(game.getCount());

        enemy.win();
        System.out.println(game.getCount());

        player.win();
        System.out.println(game.getCount());
    }
}

File: src/main/scala/sample/design_pattern/Mediator.scala
/*
 * Copyright 2013 SatoShun
 *
 * Adapter Pattern
 */

package sample.design_pattern;


class Game {
    private var count: Int = 0

    def change(player: IPlayer) = {
        if (player.isInstanceOf[Player]) {
            count += 1
        } else if (player.isInstanceOf[Enemy]) {
            count -= 1
        }
    }

    def getCount(): Int = count
}


trait IPlayer {
    def win(): Unit;
}


class Player(val game: Game) extends IPlayer {
    def win() = {
        game.change(this)
    }
}


class Enemy(val game: Game) extends IPlayer {
    def win() = {
        game.change(this)
    }
}


object Mediator {
    def main(args: Array[String]) = {
        val game: Game = new Game()
        val player: IPlayer = new Player(game)
        val enemy: IPlayer = new Enemy(game)

        player.win()
        player.win()
        println(game.getCount)

        enemy.win()
        println(game.getCount)
    }
}
